modules/fedimint-lnv2-client/src/lib.rs

```diff
@@ -0,0 +1,725 @@
+            .into_inner();
+
+        let payment_info = self.payment_info(gateway_api.clone()).await?;
+
+        ensure!(
+            max_fee >= payment_info.payment_fees.receive,
+            "The gateways receive fee is above {max_fee:?}"
+        );
+
+        let contract_amount = payment_info
+            .payment_fees
+            .incoming_contract_amount(invoice_amount.msats);
+
+        let expiration = duration_since_epoch()
+            .as_secs()
+            .saturating_add(expiry_time as u64);
+
+        let claim_pk = recipient_static_pk
+            .mul_tweak(
+                secp256k1::SECP256K1,
+                &Scalar::from_be_bytes(ephemeral_tweak).expect("Within curve order"),
+            )
+            .expect("Tweak is valid");
+
+        let contract = IncomingContract::new(
+            self.cfg.tpe_agg_pk,
+            encryption_seed,
+            preimage,
+            contract_amount,
+            expiration,
+            claim_pk,
+            payment_info.public_key,
+            ephemeral_pk,
+        );
+
+        let payload = CreateInvoicePayload {
+            federation_id: self.federation_id,
+            contract: contract.clone(),
+            invoice_amount,
+            description,
+            expiry_time,
+        };
+
+        let invoice = reqwest::Client::new()
+            .post(
+                gateway_api
+                    .join("create_invoice")
+                    .expect("'create_invoice' contains no invalid characters for a URL")
+                    .as_str(),
+            )
+            .json(&payload)
+            .send()
+            .await?
+            .json::<anyhow::Result<Bolt11Invoice, String>>()
+            .await?
+            .map_err(|error| anyhow!(error))?;
+
+        ensure!(
+            invoice.payment_hash() == &contract.commitment.payment_hash,
+            "Invoice's payment hash does not match the contract's payment hash"
+        );
+
+        ensure!(
+            invoice.amount_milli_satoshis() == Some(invoice_amount.msats),
+            "Invoice's amount does not match the requested amount"
+        );
+
+        Ok((contract, invoice))
+    }
+
+    pub async fn receive_external_contract(
+        &self,
+        contract: IncomingContract,
+    ) -> anyhow::Result<OperationId> {
+        let operation_id = OperationId(contract.consensus_hash::<sha256::Hash>().into_inner());
+
+        let (claim_keypair, agg_decryption_key) =
+            self.recover_incoming_contract_keys(&contract).await?;
+
+        let receive_sm = LightningClientStateMachines::Receive(ReceiveStateMachine {
+            operation_id,
+            contract,
+            state: ReceiveSMState::Pending(Pending {
+                claim_keypair,
+                agg_decryption_key,
+            }),
+        });
+
+        // this may only fail if the operation id is already in use, in which case we
+        // ignore the error such that the method is idempotent
```

Comment:
Instead of doing this, maybe we should just check if the operation id is already in use before recovering the incoming contract keys

Reply:
Saves us some crypto operations